src/persist-client/src/internal/compact.rs

```diff
@@ -185,67 +193,80 @@ impl Compactor {
     {
         let parts_cpu_heavy_runtime = Arc::clone(&cpu_heavy_runtime);
         let compact_blocking = async move {
-            let () = Self::validate_req(&req)?;
-
-            let mut parts = BatchParts::new(
-                cfg.batch_builder_max_outstanding_parts,
-                Arc::clone(&metrics),
-                req.shard_id,
-                writer_id,
-                req.desc.lower().clone(),
-                Arc::clone(&blob),
-                parts_cpu_heavy_runtime,
-                &metrics.compaction.batch,
-            );
+            let () = Compactor::validate_req(&req)?;
+            // compaction needs memory enough for at least 2 fetched parts + 1 in-progress part
+            assert!(memory_budget_bytes >= 3 * cfg.blob_target_size);
 
-            // TODO: Do this in a bounded amount of memory. The pattern for how
-            // to do this has already been established in old persist (see
-            // background.rs). We'll have to generalize that from 2 to N batches
-            // and also resolve the (much much harder) issue of new persist
-            // batch sorted-ness.
-            let mut updates = Vec::new();
-            for part in req.inputs.iter() {
-                for key in part.keys.iter() {
-                    fetch_batch_part(
-                        &req.shard_id,
-                        blob.as_ref(),
-                        &metrics,
-                        key,
-                        &part.desc,
-                        |k, v, mut t, d| {
-                            t.advance_by(req.desc.since().borrow());
-                            let d = D::decode(d);
-                            updates.push(((k.to_vec(), v.to_vec()), t, d));
-                        },
-                    )
-                    .await?;
-                }
-            }
-            consolidate_updates(&mut updates);
+            let mut ordered_runs = Self::order_runs::<T, D>(&req);
 
-            let len = updates.len();
-            let mut builder = ColumnarRecordsVecBuilder::new_with_len(cfg.blob_target_size);
-            for ((k, v), t, d) in updates.iter() {
-                builder.push(((k, v), T::encode(t), D::encode(d)));
+            let fetched_batch_memory_budget = memory_budget_bytes - cfg.blob_target_size;
 
-                // Flush out filled parts as we go to keep bounded memory use.
-                for chunk in builder.take_filled() {
-                    parts
-                        .write(chunk, req.desc.upper().clone(), req.desc.since().clone())
-                        .await;
+            // for now: assume each part is always blob_target_size. Given `memory_budget_bytes`
```

Comment:
let's structure this so it's easy to swap in actual sizes when we have them. something like a loop that adds sizes (initially hardcoded to cfg.blob_target_size) to some total until it's about to go over fetched_batch_memory_budget, then asserts there's at least 2 and runs the compaction on that

Reply:
👍